src/Elasticsearch/Endpoints/AbstractEndpoint.php

```diff
@@ -60,14 +64,18 @@
     /**
      * @param Transport $transport
      */
-    public function __construct($transport)
+    public function __construct($transport, RequestFactory $requestFactory = null)
     {
         $this->transport = $transport;
+        $this->requestFactory = $requestFactory;
     }
 
     /**
      * @throws \Exception
      * @return array
+     *
+     * @deprecated Performing a request directly on an endpoint is deprecated since 2.2 and will be removed in 3.0,
+     *             you should instead use the createRequest method
      */
     public function performRequest()
```

Comment:
`performRequest()` is 100% an internal method call, so it should be safe to convert all the everything to `createRequest()` internally.

However, I don't think we can remove `performRequest()` before 3.0 because a user can inject their own endpoints, and so removing `performRequest()` would break backwards compat for those custom endpoints.  The best we can do is deprecate for now.

Regarding the PRs, we can lump the refactoring into this one, or split it into smaller chunks as you/me find time to work on it.  

It might be nice to refactor one of the smaller namespaces, like [NodesNamespace](https://github.com/elastic/elasticsearch-php/blob/master/src/Elasticsearch/Namespaces/NodesNamespace.php) just to ensure the tests pass and everything works as expected?  Then the rest we can do as time allows?

Note: I'm going to be pretty busy for the next week or two, need to prepare my talks for Elasticon :)